File: drivers/zig/driver.zig
// Crucible Zig driver — persistent replay front-end for the differential loop.
//
// Speaks drivers/common/CONTRACT.md: reads length-prefixed records on stdin,
// decodes each into the probe message via the generated corelib-zig code, and
// writes one canonical line (oracle/canonical.md) per record to stdout.
//
// Like Go/Python/Java/TS/C# (and unlike Rust/C++), the generated Zig decode is
// fallible: `Probe.decode` returns `sofab.Error!Probe`, so the verdict is a
// `catch`. Zig 0.16 std.Io: main takes `std.process.Init` (for `io`/`gpa`),
// stdin/stdout go through `std.Io.File` reader/writer interfaces.
//
// The generated decode is zero-copy: `m.s` borrows from `data`, so the canonical
// line is emitted while `data` is still alive (before it is freed).
const std = @import("std");
const message = @import("message.zig");
const sofab = @import("sofab");

// --- materialized value dump (oracle/materialized.md), SOFAB_MATERIALIZE=1 ------
//
// The default accept path re-encodes the decoded value to wire (schema-agnostic,
// but blind to a decode that differs only where the sparse-canonical wire elides —
// see canonical.md §Tradeoff). This second path instead walks the decoded Probe
// value directly and dumps every field + every array element explicitly, matching
// engine/structured/materialize.py byte-for-byte.
//
// The walker is SCHEMA-AGNOSTIC: it is generated at build time from the descriptor
// oracle/materialized-schema.json by materialize_gen.py (run by build.sh), which
// unrolls the descriptor into straight-line field-access code. A schema change
// reshapes materialize_gen.zig with zero hand-editing here. (Zig 0.16 comptime
// field access needs field names at compile time, and string vs blob are both
// []const u8 — so the descriptor is unrolled to source rather than walked at
// runtime.) Strings/blobs borrow `data` (zero-copy), so the dump is written while
// `data` is still alive (before reset).
const matgen = @import("materialize_gen.zig");

// ---- the streaming axes (drivers/common/CONTRACT.md) --------------------------
//
// The replay protocol hands each record over whole and re-encodes it with one call,
// so neither streaming surface of the generated API is reachable through it. Unset,
// every variable below is today's behaviour byte for byte.
//
// Zig's generated Decoder exposes status(), so the verdict comes from there rather
// than from finish() (which returns an error mid-field).
//
// SOFAB_CHUNK_SCRUB is NOT APPLICABLE here, and that is documented backend
// behaviour rather than a defect. The generated Decoder states it outright: "a
// string or blob that arrives whole inside one chunk is borrowed from that chunk
// ... so a fed chunk must outlive the message." Scrubbing a chunk after feed
// therefore violates this backend's contract instead of testing it, and the driver
// exits 3 (the same "not applicable at this setting" code the encode axis uses for
// an unusable SOFAB_FLUSH) rather than manufacturing a mismatch. That the
// chunk-lifetime contract differs across backends at all — corelib-cpp, -rs, -c-cpp
// and the managed runtimes all copy — is worth a family-level answer; recorded in
// docs/TODO.md.
const StreamCfg = struct {
    split: usize = 0,
    chunk: usize = 0,
    enc_stream: bool = false,
    flush: usize = 0,
};

fn envUsize(init: std.process.Init, name: []const u8) usize {
    const v = init.environ_map.get(name) orelse return 0;
    return std.fmt.parseInt(usize, v, 10) catch 0;
}

fn readStreamCfg(init: std.process.Init) StreamCfg {
    var cfg = StreamCfg{};
    cfg.split = envUsize(init, "SOFAB_SPLIT");
    cfg.chunk = envUsize(init, "SOFAB_CHUNK");
    cfg.flush = envUsize(init, "SOFAB_FLUSH");
    if (init.environ_map.get("SOFAB_CHUNK_SCRUB")) |v| {
        if (v.len != 0 and !std.mem.eql(u8, v, "0")) {
            std.debug.print("crucible-zig: SOFAB_CHUNK_SCRUB is not applicable — this " ++
                "backend borrows string/blob payloads that arrive whole in one chunk, " ++
                "and documents that a fed chunk must outlive the message\n", .{});
            std.process.exit(3);
        }
    }
    if (init.environ_map.get("SOFAB_ENCODE")) |e| {
        if (std.mem.eql(u8, e, "stream")) {
            cfg.enc_stream = true;
        } else if (std.mem.eql(u8, e, "to")) {
            std.debug.print("crucible-zig: SOFAB_ENCODE=to — this backend has no " ++
                "encodeTo (it has new, stream)\n", .{});
            std.process.exit(2);
        } else if (e.len != 0 and !std.mem.eql(u8, e, "new")) {
            std.debug.print("crucible-zig: unknown SOFAB_ENCODE (this backend has " ++
                "new, stream)\n", .{});
            std.process.exit(2);
        }
    }
    // Announce on stderr (never parsed). A driver that silently ignored these would
    // be indistinguishable from one that honours them — stdout is identical either
    // way — so this makes "it really re-feeds" checkable rather than asserted.
    if (cfg.split != 0 or cfg.chunk != 0 or cfg.flush != 0 or cfg.enc_stream) {
        std.debug.print("crucible-zig: streaming cfg split={d} chunk={d} enc={s} " ++
            "flush={d}\n", .{
            cfg.split, cfg.chunk,
            if (cfg.enc_stream) @as([]const u8, "stream") else @as([]const u8, "new"),
            cfg.flush,
        });
    }
    return cfg;
}

// Sink for the streaming encode: collect what the OStream flushes.
const EncAcc = struct {
    list: std.ArrayListUnmanaged(u8) = .empty,
    alloc: std.mem.Allocator,

    fn sink(ctx: ?*anyopaque, data: []const u8) void {
        const self: *EncAcc = @ptrCast(@alignCast(ctx.?));
        self.list.appendSlice(self.alloc, data) catch {};
    }
};

pub fn main(init: std.process.Init) !void {
    const io = init.io;

    // SOFAB_MATERIALIZE=1 selects the materialized-value dump on accept
    // (oracle/materialized.md); unset keeps the default round-trip hex path.
    const materialize_mode = if (init.environ_map.get("SOFAB_MATERIALIZE")) |v|
        std.mem.eql(u8, v, "1")
    else
        false;

    const cfg = readStreamCfg(init);
    const chunking = cfg.split != 0 or cfg.chunk != 0;

    var inbuf: [8192]u8 = undefined;
    var stdin_reader = std.Io.File.stdin().readerStreaming(io, &inbuf);
    const in = &stdin_reader.interface;

    var outbuf: [8192]u8 = undefined;
    var stdout_writer = std.Io.File.stdout().writer(io, &outbuf);
    const out = &stdout_writer.interface;

    // Per-record arena: the full message decodes array storage from it and
    // re-encode allocates from it; reset per record so nothing leaks across the
    // (potentially millions of) inputs.
    var arena = std.heap.ArenaAllocator.init(init.gpa);
    defer arena.deinit();

    while (true) {
        const lenb = in.takeArray(4) catch |e| switch (e) {
            error.EndOfStream => break, // clean EOF at record boundary
            else => return e,
        };
        const n: usize = @as(usize, lenb[0]) | (@as(usize, lenb[1]) << 8) |
            (@as(usize, lenb[2]) << 16) | (@as(usize, lenb[3]) << 24);

        _ = arena.reset(.retain_capacity);
        const a = arena.allocator();

        const data = try a.alloc(u8, n);
        if (n > 0) try in.readSliceAll(data);

        // decode -> re-encode -> hex (oracle/canonical.md). m borrows string bytes
        // from `data` (kept alive until the next reset), so encode can read them.
        //
        // The chunked path is taken ONLY when a chunking variable is set, so the
        // default stays the one-shot Probe.decode byte for byte — which is what
        // makes the gate meaningful: it then compares two genuinely different code
        // paths. `data` outlives the message either way (arena reset per record),
        // which is exactly the lifetime this backend documents as required.
        const m = (if (chunking) blk: {
            var acc: message.Probe = .{};
            var d = message.Probe.decoder(&acc, a);
            var off: usize = 0;
            while (off < n) {
                var step: usize = if (cfg.chunk > 0) cfg.chunk else n;
                if (cfg.chunk == 0 and cfg.split > 0 and cfg.split < n and off == 0)
                    step = cfg.split;
                if (off + step > n) step = n - off;
                _ = d.feed(data[off .. off + step]) catch |e| break :blk @as(
                    message.DecodeError!message.Probe, e);
                off += step;
            }
            // Verdict from status(), never from finish() (CONTRACT.md).
            if (d.status() == .incomplete)
                break :blk @as(message.DecodeError!message.Probe, error.IncompleteMessage);
            break :blk @as(message.DecodeError!message.Probe, acc);
        } else message.Probe.decode(a, data)) catch |err| {
            // INCOMPLETE (§7) is a distinct verdict, not an error: the bytes end
            // inside a field/varint or an open sequence. Emit `I` — never collapse
            // it into A (accept-as-done) or R (reject-as-malformed).
            if (err == error.IncompleteMessage) {
                try out.writeAll("I\n");
                try out.flush();
                continue;
            }
            if (err == error.LimitExceeded) {
                // LIMIT_EXCEEDED (generator#102, limit mode only): a configured receiver-side
                // cap on a schema-unbounded field. A policy rejection distinct from INVALID —
                // its own verdict `L`, not `R`.
                try out.writeAll("L\n");
                try out.flush();
                continue;
            }
            const cls = switch (err) {
                error.InvalidMessage => "invalid_msg",
                error.InvalidArgument => "argument",
                // error.UsageError was removed in corelib-zig#23; the canonical
                // "usage" class stays in oracle/canonical.md for the corelibs that
                // still have it. This switch is exhaustive over the error set, so
                // the arm has to go rather than fall through.
                error.BufferFull => "buffer_full",
                error.IncompleteMessage => unreachable, // handled above
                error.LimitExceeded => unreachable, // handled above
            };
            try out.print("R {s}\n", .{cls});
            try out.flush();
            continue;
        };
        // Accept. In materialize mode (oracle/materialized.md) dump the decoded
        // value's fields/elements explicitly instead of re-encoding to wire. m
        // borrows string/blob bytes from `data` (alive until the next reset), so
        // the whole dump is written now, before the loop resets the arena.
        if (materialize_mode) {
            try out.writeAll("A ");
            try matgen.materialize(out, &m);
            try out.writeAll("\n");
            try out.flush();
            continue;
        }

        // Re-encode through the surface SOFAB_ENCODE selects. Both must emit
        // identical bytes, and SOFAB_FLUSH must not change them either: it gives the
        // OStream an n-byte buffer with a sink, so the encoder crosses a buffer
        // boundary at every offset — the encode-side mirror of SOFAB_CHUNK=1.
        const enc = (if (cfg.enc_stream) blk: {
            const cap = if (cfg.flush > 0) cfg.flush else message.Probe.MAX_SIZE;
            const sbuf = a.alloc(u8, cap) catch break :blk error.BufferFull;
            var acc = EncAcc{ .alloc = a };
            var os = sofab.OStream.initFlush(sbuf, 0, &acc, EncAcc.sink);
            m.serialize(&os) catch break :blk error.BufferFull;
            _ = os.flush();
            break :blk @as(anyerror![]u8, acc.list.items);
        } else m.encode(a)) catch {
            try out.writeAll("R other\n");
            try out.flush();
            continue;
        };
        try out.writeAll("A ");
        for (enc) |b| try out.print("{x:0>2}", .{b});
        try out.writeAll("\n");
        try out.flush();
    }
}

File: drivers/rust/driver.rs
// Crucible Rust driver — shared body for BOTH corelibs (corelib-rs / corelib-rs-no-std).
//
// build.sh prepends a per-variant preamble that brings `Probe` into scope:
//   std     : `mod message; use message::Probe;`
//   no-std  : `use sofabuffers_generated::Probe;`
// The `sofab` crate (either corelib) is a dependency of both, so the imports
// below resolve unchanged.
//
// Single-pass decode via the generated fallible `Probe::try_decode` (sofabgen
// 0.16.0 — G-0001 fixed, see docs/SOFABGEN.md): it runs the real generated visitor
// AND returns the §7 outcome as a `Result`, so one call yields both the verdict
// and the value. This replaced the earlier two-pass workaround, which recovered
// the verdict by re-running `IStream::feed` against a null visitor because the old
// infallible `Probe::decode` discarded feed's Result. Because the null visitor
// skipped the generated per-field checks, that workaround also missed the
// over-count-array rejection (generator#100); try_decode runs them, so rust now
// converges with the family on those inputs (was the F-0003 residual divergence).
//
// LimitExceeded (generator#102) maps to a fourth verdict `L`, gated behind the
// `limit` cargo feature: the arm is std-only (corelib-rs-no-std's Error has no
// LimitExceeded variant), so build.sh enables `limit` for the `rs` variant only.
// It fires solely under a configured cap (limit mode); with no cap it never occurs,
// so the default conformance run is unchanged.
//
// Emits the canonical form (oracle/canonical.md) over the replay protocol
// (drivers/common/CONTRACT.md).
use sofab::Error;
use std::io::{Read, Write};

// ---- the streaming axes (drivers/common/CONTRACT.md, "The streaming axes") -------
//
// The replay protocol hands each record over whole and re-encodes it with one call, so
// neither streaming surface of the generated API is reachable through it. Unset, every
// variable below is today's behaviour byte for byte.
//
// Rust's generated `Decoder` has **no `status`** — crucible#132's API table says every
// chunked decoder exposes one, and at sofabgen cfe5250b this backend does not. So the
// verdict comes from `finish()`, which is sound here for the reason the contract gives:
// `finish()` returns `Result<Probe, sofab::Error>`, the *same* three-valued outcome
// `try_decode` returns, so routing through it introduces no API-shape difference. It is
// also more than a formality — `finish()` feeds an empty chunk to probe end-of-input,
// which is exactly what makes a truncated stream an error rather than a half-filled
// value.
//
// This backend has no `encodeTo`, so `SOFAB_ENCODE=to` is a hard error rather than a
// fallback (meta: encode_surfaces=new,stream).
#[derive(Clone, Copy, PartialEq)]
enum EncSurface {
    New,
    Stream,
}

struct StreamCfg {
    split: usize,
    chunk: usize,
    scrub: bool,
    enc: EncSurface,
    flush: usize,
}

fn env_usize(name: &str) -> usize {
    std::env::var(name).ok().and_then(|v| v.parse().ok()).unwrap_or(0)
}

fn read_stream_cfg() -> StreamCfg {
    let enc = match std::env::var("SOFAB_ENCODE").as_deref() {
        Ok("to") => {
            eprintln!(
                "crucible-rust: SOFAB_ENCODE=to — this backend has no encodeTo \
                 (it has new, stream)"
            );
            std::process::exit(2);
        }
        Ok("stream") => EncSurface::Stream,
        Ok("new") | Err(_) => EncSurface::New,
        Ok("") => EncSurface::New,
        Ok(other) => {
            eprintln!(
                "crucible-rust: unknown SOFAB_ENCODE={other} (this backend has new, stream)"
            );
            std::process::exit(2);
        }
    };
    let cfg = StreamCfg {
        split: env_usize("SOFAB_SPLIT"),
        chunk: env_usize("SOFAB_CHUNK"),
        scrub: matches!(std::env::var("SOFAB_CHUNK_SCRUB").as_deref(), Ok(v) if !v.is_empty() && v != "0"),
        enc,
        flush: env_usize("SOFAB_FLUSH"),
    };
    // Announce on stderr (never parsed). A driver that silently ignored these would be
    // indistinguishable from one that honours them — identical stdout either way — so
    // this is what makes "it really re-feeds" checkable rather than asserted.
    if cfg.split != 0 || cfg.chunk != 0 || cfg.scrub || cfg.flush != 0
        || cfg.enc != EncSurface::New
    {
        eprintln!(
            "crucible-rust: streaming cfg split={} chunk={} scrub={} enc={} flush={}",
            cfg.split,
            cfg.chunk,
            cfg.scrub as u8,
            if cfg.enc == EncSurface::New { "new" } else { "stream" },
            cfg.flush
        );
    }
    cfg
}

// How the record is cut on its way into the decoder. Never an empty chunk: k<=0,
// k>=len and n>=len all mean one chunk holding the whole record. A zero-length record
// yields no chunks at all.
fn slices(cfg: &StreamCfg, len: usize) -> Vec<(usize, usize)> {
    if len == 0 {
        return Vec::new();
    }
    if cfg.chunk > 0 {
        return (0..len)
            .step_by(cfg.chunk)
            .map(|o| (o, std::cmp::min(cfg.chunk, len - o)))
            .collect();
    }
    if cfg.split > 0 && cfg.split < len {
        return vec![(0, cfg.split), (cfg.split, len - cfg.split)];
    }
    vec![(0, len)]
}

// Feed the record in the configured pieces through ONE decoder, then take the verdict
// from `finish()`. An `Incomplete` from a mid-stream `feed` is not terminal — it says
// only that *those bytes* ended mid-field; any other error is.
fn decode_streamed(cfg: &StreamCfg, data: &[u8]) -> Result<Probe, Error> {
    let mut d = Probe::decoder();
    let mut scratch: Vec<u8> = Vec::new();
    for (off, n) in slices(cfg, data.len()) {
        let r = if cfg.scrub {
            // Scrub needs a buffer the driver owns: feed it, then overwrite. A decoder
            // that borrowed from the chunk rather than copying out of it reads 0xA5.
            scratch.clear();
            scratch.extend_from_slice(&data[off..off + n]);
            let r = d.feed(&scratch);
            scratch.fill(0xA5);
            r
        } else {
            d.feed(&data[off..off + n])
        };
        match r {
            Ok(()) => {}
            Err(Error::Incomplete) => {}      // mid-field between chunks: expected
            Err(e) => return Err(e),          // terminal
        }
    }
    d.finish()
}

// Which generated call produces the `A <hex>` payload. Both surfaces must emit
// identical bytes for one decoded value, and SOFAB_FLUSH must not change that: it hands
// the OStream an n-byte buffer with a sink, so the encoder crosses a buffer boundary at
// every offset — the encode-side mirror of SOFAB_CHUNK=1.
fn encode_via(cfg: &StreamCfg, m: &Probe) -> Vec<u8> {
    if cfg.enc == EncSurface::New {
        // `encode()` is Vec<u8> on std and heapless::Vec<u8, MAX_SIZE> on no-std;
        // iterate rather than convert, so this one line compiles for both.
        return m.encode().iter().copied().collect();
    }
    let cap = if cfg.flush > 0 { cfg.flush } else { Probe::MAX_SIZE };
    let mut buf = vec![0u8; std::cmp::max(cap, 1)];
    let mut out: Vec<u8> = Vec::new();
    {
        let sink = |data: &[u8]| out.extend_from_slice(data);
        // `with_flush` is fallible since corelib-rs#86 (it always was on
        // corelib-rs-no-std): it enforces the CORELIB_PLAN §5.1 streaming minimum,
        // `buflen - offset >= MIN_OUTPUT_BUFFER`, and refuses a smaller buffer at the
        // installation. Report that refusal as exit 3 — "this backend cannot operate at
        // this configuration" — rather than unwrapping, so the oracle sees the buffer
        // size the port declined instead of a panic.
        let mut os = match sofab::OStream::with_flush(&mut buf, 0, sink) {
            Ok(os) => os,
            Err(e) => {
                eprintln!(
                    "crucible-rust: OStream::with_flush refused a {}-byte buffer \
                     (MIN_OUTPUT_BUFFER={}): {:?}",
                    buf.len(),
                    sofab::MIN_OUTPUT_BUFFER,
                    e
                );
                std::process::exit(3);
            }
        };
        m.serialize(&mut os);
        os.flush();
    }
    out
}

fn reject_class(e: Error) -> &'static str {
    match e {
        Error::InvalidMsg => "invalid_msg",
        Error::Argument => "argument",
        // Error::Usage was removed in corelib-rs#35 / corelib-rs-no-std#55; the
        // canonical "usage" class stays in oracle/canonical.md for the corelibs
        // that still have it. Anything unmapped falls through to "other".
        Error::BufferFull => "buffer_full",
        _ => "other", // Error is #[non_exhaustive]
    }
}

// ---- materialized value dump (oracle/materialized.md), SOFAB_MATERIALIZE=1 -------
//
// The default accept path re-encodes the decoded value to wire (schema-agnostic,
// but blind to a decode that differs only where the sparse-canonical wire elides —
// canonical.md §Tradeoff). This path instead walks the DECODED value and dumps every
// field + every array element explicitly, matching engine/structured/materialize.py
// byte-for-byte. The value is already faithful (the decoder fills numeric/fp arrays
// to N, grows the wrapper Vecs to highest-populated-index + 1, and omitted scalar
// fp fields decode to their +0.0 default), so we dump it as-is with no normalization.
//
// The walker itself is NOT hand-written: `materialize_gen.py` unrolls the schema
// descriptor (oracle/materialized-schema.json) into straight-line field-access code
// at build time, and build.sh drops it beside this file as `materialize_gen.rs`
// (Rust has no runtime reflection, so a runtime table cannot drive it — the source is
// generated instead). A schema change regenerates the walker with zero edits here.
//
// The generated `pub fn materialize(m: &Probe) -> String` compiles for BOTH corelibs:
// it touches only member APIs shared by the std and no_std container flavors
// (`.as_bytes()` on strings, slice-deref `&x[..]` on blobs, `.iter()` over the
// wrappers — the numeric/fp scalar and array fields are identical in both) and builds
// its output with `core::fmt::Write` into a `String` (the driver binary is always std
// for both corelib variants). We then write those bytes to the std::io sink exactly as
// the round-trip path does.
include!("materialize_gen.rs");

fn canonical(out: &mut impl Write, data: &[u8], materialize_mode: bool, cfg: &StreamCfg) {
    // The chunked path is taken ONLY when a chunking variable is set, so the default
    // stays the one-shot try_decode byte for byte — which is also what makes the gate
    // meaningful: it then compares two genuinely different code paths.
    let decoded = if cfg.split != 0 || cfg.chunk != 0 || cfg.scrub {
        decode_streamed(cfg, data)
    } else {
        Probe::try_decode(data)
    };
    match decoded {
        Ok(m) => {
            if materialize_mode {
                // COMPLETE, materialize mode: dump the decoded value (materialized.md).
                let _ = write!(out, "A ");
                let _ = out.write_all(materialize(&m).as_bytes());
                let _ = writeln!(out);
                return;
            }
            // COMPLETE: re-encode the decoded value -> hex, through whichever encode
            // surface SOFAB_ENCODE selects (default: the allocating encode()).
            let bytes = encode_via(cfg, &m);
            let _ = write!(out, "A ");
            for b in bytes.iter() {
                let _ = write!(out, "{:02x}", b);
            }
            let _ = writeln!(out);
        }
        Err(Error::Incomplete) => {
            // INCOMPLETE (MESSAGE_SPEC §7): the bytes end mid-message — the third
            // canonical verdict, neither accept (A) nor reject (R). Not an error.
            let _ = writeln!(out, "I");
        }
        #[cfg(feature = "limit")]
        Err(Error::LimitExceeded) => {
            // LIMIT_EXCEEDED (generator#102, limit mode only): a configured
            // receiver-side cap on a schema-unbounded field was exceeded. A policy
            // rejection distinct from INVALID — its own verdict `L`, not `R`.
            let _ = writeln!(out, "L");
        }
        Err(e) => {
            let _ = writeln!(out, "R {}", reject_class(e));
        }
    }
}

fn main() {
    // Materialize mode (oracle/materialized.md): on a COMPLETE decode, emit a value
    // dump instead of the re-encoded wire hex. Read once at startup; every other
    // verdict path is unaffected. The driver binary is std for both corelib variants,
    // so std::env is available under the no_std corelib too.
    let materialize_mode = std::env::var("SOFAB_MATERIALIZE").as_deref() == Ok("1");
    let cfg = read_stream_cfg();

    let stdin = std::io::stdin();
    let mut r = stdin.lock();
    let stdout = std::io::stdout();
    let mut w = std::io::BufWriter::new(stdout.lock());

    let mut lenbuf = [0u8; 4];
    loop {
        match r.read_exact(&mut lenbuf) {
            Err(ref e) if e.kind() == std::io::ErrorKind::UnexpectedEof => break, // clean EOF
            Err(e) => {
                eprintln!("crucible-rust: short length prefix: {e}");
                std::process::exit(1);
            }
            Ok(()) => {}
        }
        let n = u32::from_le_bytes(lenbuf) as usize;
        let mut data = vec![0u8; n];
        if n > 0 {
            if let Err(e) = r.read_exact(&mut data) {
                eprintln!("crucible-rust: short payload: {e}");
                std::process::exit(1);
            }
        }
        canonical(&mut w, &data, materialize_mode, &cfg);
        w.flush().ok();
    }
}
